torch_xla/experimental/pjrt.py

```diff
@@ -220,6 +233,49 @@ def _discover_tpu_master_worker_ip(device: torch.device):
   return _merge_replica_results(replica_results)
 
 
+@requires_pjrt
+def _run_singleprocess(fn: Callable[..., R],
+                       *args,
+                       start_method: str = 'spawn',
+                       master_port: int = 12355,
+                       **kwargs) -> Dict[int, R]:
+  """Runs `fn` on a single device core.
+
+  Spawns one process on a single physical device (e.g. TPU chip).
+
+  Args:
+    fn: Function to run on the device devices
+    args: args to pass to `fn`
+    start_method: The Python `multiprocessing` process creation method.
+      Default: `spawn`
+    kwargs: kwargs to pass to `fn`
+
+  Returns:
+    the result of calling `fn`.
+  """
+  with concurrent.futures.ProcessPoolExecutor(
+      max_workers=1,
+      mp_context=torch.multiprocessing.get_context(start_method)) as executor:
+    os.environ.setdefault('LOCAL_WORLD_SIZE', '1')
+
+    if device_type() == 'TPU':
+      tpu.configure_topology(0, 1)
+
+    devices = xm.get_xla_supported_devices()[:1]
+    xm.set_replication(xm.xla_device(), [])
+
+    @functools.wraps(fn)
+    def _thread_fn(device: torch.device):
+      torch_xla._XLAC._xla_set_default_device(devices[0])
+
+      return fn()
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
```

Comment:
You don't need to create a thread pool executor here, since we're only running one instance of `fn`.

The only reason you would need a thread pool that I see is for `_maybe_init_torch_distributed`. However, the only reason we need a thread pool there is we need to have to use a hack to discover the TPU master IP. We actually don't need to contact the master over TCP at all to initialize torch.distributed in this case, since there's only one process.

There's a default in `init_pjrt_process_group` to use a random file to initialize when in a single process context. So, you can actually just directly call `init_pjrt_process_group` here and remove the `ThreadPoolExecutor` entirely.

Reply:
Cool! Done.